[donwong2308] iP

donwong2308 opened the pull request:
### PeWPro

> "Never memorize something that you can look up.” – Albert Einstien ([source](https://parade.com/1240718/kelseypelzer/albert-einstein-quotes/))

PeWPro frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use 👍 

GFMD elements:

- [x] a heading
- [x] a bullet list
- [x] a numbered list
- [x] a fenced code block (with syntax highlighting)
- [x] a task list
- [x] an emoji
- [x] a blockquote
- [x] a hyperlink
- [x] inline code
- [x] some text formatting: bold, italic, strikethrough etc.

Features:

1. Add Tasks
2. Manage deadlines
3. Delete Tasks
4. Reminder Deadlines (Coming Soon)

If you are a Java programmer, you can use it to practice Java too. Here's the ```main``` method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

donwong2308 committed:
Did Level 0

donwong2308 committed:
Did Level 1

donwong2308 committed:
Did Level 2

donwong2308 committed:
Do Level 3

added marking an unmarking ability with a try catch for invalid inputs

donwong2308 committed:
Do Level-4

added todos, deadlines and events as task types

donwong2308 committed:
Do Level-5

Added 3 exception handling, 2 from the guide and 1 for checking array index validity when marking/unmarking

donwong2308 committed:
Do Level-6

Added delete function

donwong2308 committed:
Created Tests for automate testing

added inputs and expected outputs

donwong2308 committed:
added saving functionality

able to read tasks from text file and save back to text file

donwong2308 committed:
Merge pull request #1 from donwong2308/branch-Level-7

Level-7

donwong2308 committed:
Finished converting dates to datetime

finished the date time format, added exception handling for invalid inputs

donwong2308 committed:
Merge pull request #2 from donwong2308/branch-Level-8

Finished converting dates to datetime

donwong2308 committed:
Add more OOP

Created 4 extra classes,
ui, tasklist, storage and parser

donwong2308 committed:
Merge pull request #3 from donwong2308/A-MoreOOP

Add more OOP

donwong2308 committed:
Divide classes into packages

Placed all classes in one package

donwong2308 committed:
Merge pull request #4 from donwong2308/branch-A-Packages

Divide classes into packages

donwong2308 committed:
Merge pull request #5 from donwong2308/add-gradle-support

Add gradle support

donwong2308 committed:
Add JUnit Tests

Added 2 tests for todo get task and
deadline date validity

donwong2308 committed:
Merge pull request #6 from donwong2308/branch-A-JUnit

Add JUnit Tests

donwong2308 committed:
Add JavaDocs

Added javadocs for todo, deadline, event, task, duke, parser and dukeexception

donwong2308 committed:
Merge pull request #7 from donwong2308/branch-A-JavaDoc

Add JavaDocs

donwong2308 committed:
Add Level-9 and change tasklist type

Added the find function and changed tasklist type from array to tasklist

donwong2308 committed:
Merge pull request #8 from donwong2308/branch-Level-9

Add Level-9 and change tasklist type

donwong2308 committed:
Edit code syntax

Did coding standard

donwong2308 committed:
Merge pull request #9 from donwong2308/branch-A-CodingStandard

Edit code syntax

donwong2308 committed:
Merge pull request #10 from donwong2308/add-gradle-support

Add gradle support

tituschewxj commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,80 @@
+    /**
+     * Constructs a Duke object
+     * @param filePath the path to the file containing the tasks
+     */
```
Perhaps all your Javadoc comments can have an empty line between the description and parameter section?

tituschewxj commented on src/test/java/duke/TodoTest.java:
```diff
@@ -0,0 +1,35 @@
+                    "d-M-yyyy[ HHmm]"
+            };
+            assertEquals("Invalid date and time format, use the following formats: " +
+                    "\n" + Arrays.toString(dateFormats), d.getMessage());
```
```suggestion
assertEquals("Invalid date and time format, use the following formats: \n"
                    + Arrays.toString(dateFormats), d.getMessage());
```

tituschewxj commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,59 @@
+    }
+
+    public String getStatusIcon() {
+        return (isDone ? "X" : " "); // mark done task with X
```
Perhaps this is a more acceptable way to format a ternary expression?
```suggestion
        return (isDone) ? "X" : " "; // mark done task with X
```

tituschewxj commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,30 @@
+
+    @Override
+    public String getTask() {
+        return index + ". [" + type + "][" + getStatusIcon() + "] " + description + "(from: " + start + " to: " + end + ")";
```
Should your line length be longer than 120 chars?
```suggestion
        return index + ". [" + type + "][" + getStatusIcon() + "] "
                + description + "(from: " + start + " to: " + end + ")";
```

tituschewxj commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,153 @@
+                } catch (DukeException d) {
+                    System.out.println(d);
+                }
+            }
```
I didn't like how nested your code is. Perhaps refactor it out to multiple methods?

tituschewxj commented on src/main/java/duke/DukeException.java:
```diff
@@ -0,0 +1,90 @@
+     * @param task_arr list of tasks
+     * @throws DukeException if the index is invalid
+     */
+    static void validateArrIndex(int index, ArrayList<Task> task_arr) throws DukeException {
```
Perhaps `task_arr` can be in camelCase? I noticed the same issue in several other places too.

tituschewxj commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,153 @@
+                }
+            }
+        } catch (DukeException d) {
+            System.out.println( d);
```
```suggestion
            System.out.println(d);
```

tituschewxj reviewed (commented):
Overall, I found your code easy to read for the most part, except a few places where the nesting was too deep. I noticed a few minor coding standard violations too. Some of the methods are getting quite long. Consider splitting into smaller methods if that makes sense.

ryanlimdx commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,80 @@
+        return scanner.nextLine();
+    }
+
+    private void saveTasksToFile() {
+        try {
+            FileWriter writer = new FileWriter("data/duke.txt");
+            for (Task task : taskList.getTaskArr()) {
+                writer.write(task.save() + "\n");
+            }
+            writer.close();
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+    }
```
Perhaps it could be better to extract this method out to Storage class instead? This way data can be read and stored from the storage class, making it cleaner.

ryanlimdx commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,62 @@
+        return task_arr;
+    }
+
+    // Add other methods for saving tasks if needed
```
As mentioned in the other review; I think it would be good to collate read write tasks in the Storage class here!

ryanlimdx commented on src/main/java/duke/TaskList.java:
I like the naming in this class. It makes it very understandable what is the method for!

ryanlimdx commented on src/main/java/duke/Parser.java:
This class is running a little long, perhaps some parts could be extracted such as `formatDateTime` (along with other date time methods and exceptions) into a datetime class?

ryanlimdx commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,27 @@
+package duke;
+
+public class Deadline extends Task {
+    protected String type = "D";
```
Good use taking out the type and representing it in a variable!

ryanlimdx commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,59 @@
+     * saves the task as a string
+     * @return the string representation of the task
+     */
+    public String save() {
```
Perhaps naming could be changed to `saveAsString` to distinguish between that and saving to the file?

ryanlimdx reviewed (commented):
Overall, looks good to me! Most others have been pointed out by @tituschewxj. Some classes could be shorter by abstracting out methods, but it's pretty minor.

donwong2308 committed:
Add GUI

Followed JavaFX tutorial to do a simple GUI

donwong2308 committed:
Merge pull request #11 from donwong2308/branch-Level-10

Merge pull request #10 from donwong2308/branch-Level-10

donwong2308 committed:
Modify text

donwong2308 committed:
Add GUI

donwong2308 committed:
Update gitignore

Made git to ignore the data folder

donwong2308 committed:
Add Assertions and Enums

Added assertion to check for valid instruction type.

Added enums class since instruction type is fixed and can be its own datatype

donwong2308 committed:
Merge pull request #12 from donwong2308/branch-A-Assertions

Add Assertions and Enums

donwong2308 committed:
Update code quality

Made parser use switch statements to improve readability.

donwong2308 committed:
Merge pull request #13 from donwong2308/branch-A-CodeQuality

Update code quality

donwong2308 committed:
Update product name

Renamed some methods and variables from Duke to Pew

donwong2308 committed:
Update TaskList.java

Uses .contains method to find tasks based on keyword

donwong2308 committed:
Merge pull request #15 from donwong2308/C-BetterSearch

Add C-BetterSearch extension

donwong2308 committed:
Fix Jar File Creation

Modified mainclass in build gradle file to create working jar file

donwong2308 committed:
Add unit test

Added 1 more JUnit test and update code quality

donwong2308 committed:
Update exit

Update the application to exit when type bye

donwong2308 committed:
Update file path for final release

Change file path to just pew.txt

donwong2308 committed:
Merge pull request #16 from donwong2308/branch-A-Release

Update file path for final release

donwong2308 committed:
Update README.md

donwong2308 committed:
Update exceptions

Update array index exception handling

donwong2308 committed:
Update Exception

Updated exception handling for deadline and events

donwong2308 committed:
Clean code quality

Shorten long methods in storage and parser

donwong2308 committed:
Update README.md
